### PR TITLE
Refactor Env for better readability

### LINKED ISSUE #4240: Refactor Env for better readability
Change method signatures:

Env:
| Before | After |
|------- |  ------|
| Status is_directory(const std::string&, bool*) | StatusOr\<bool\> is_directory(const std::string&) |
| Status get_file_size(const std::string&, uint64_t*)  | StatusOr\<int64_t\> get_file_size(const std::string&) |
| Status get_file_modified_time(const std::string&, uint64_t*) | StatusOr\<uint64_t\> get_file_modified_time(const std::string&) |
| Status path_exists(const std::string&) | StatusOr<bool>  path_exists(const std::string&) |

RandomAccessFile:
| Before | After |
|--------|------|
| Status size(uint64_t*) const | StatusOr\<int64_t\> get_size() |

RandomRWFile:
| Before | After |
|--------|------|
| Status size(uint64_t*) const | StatusOr\<int64_t\> get_size() |

### PR DESCRIPTION
**Pull request by xlwh**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #4240

### CODE REVIEW

File: be/src/env/env_broker.cpp [hunk 1]
```diff
@@ -469,29 +462,27 @@ Status EnvBroker::sync_dir(const std::string& dirname) {
...
     return Status::NotSupported("BrokerEnv::canonicalize");
 }
 
-Status EnvBroker::_get_file_size(const std::string& path, uint64_t* size) {
+StatusOr<bool> EnvBroker::_get_file_size(const std::string& path) {
     TBrokerFileStatus stat;
     Status st = _list_file(path, &stat);
-    *size = stat.size;
-    return st;
+    return stat.size;
```
Comment (trueeyu) [inline be/src/env/env_broker.cpp:478]: Size should be set regardless of success or fail, should add another macro definition to cover it;

File: be/src/env/env_broker.cpp [hunk 2]
```diff
@@ -469,29 +462,27 @@ Status EnvBroker::sync_dir(const std::string& dirname) {
...
-    return Status::OK();
+    return stat.isDir;
 }
 
 Status EnvBroker::canonicalize(const std::string& path, std::string* file) {
     return Status::NotSupported("BrokerEnv::canonicalize");
 }
 
-Status EnvBroker::_get_file_size(const std::string& path, uint64_t* size) {
+StatusOr<bool> EnvBroker::_get_file_size(const std::string& path) {
```
Comment (sduzh) [inline be/src/env/env_broker.cpp:475]:
```suggestion
StatusOr<uint64_t> EnvBroker::_get_file_size(const std::string& path) {
```

File: be/src/env/env_broker.cpp [hunk 3]
```diff
@@ -469,29 +462,27 @@ Status EnvBroker::sync_dir(const std::string& dirname) {
...
     TBrokerFileStatus stat;
     Status st = _list_file(path, &stat);
```
Comment (sduzh) [inline be/src/env/env_broker.cpp:477]: RETURN_IF_ERROR(_list_file(path, &stat));

File: be/src/env/env_memory.cpp [hunk 4]
```diff
@@ -91,8 +87,7 @@ class MemorySequentialFile final : public SequentialFile {
     const std::string& filename() const override { return _random_file.filename(); }
 
     Status skip(uint64_t n) override {
-        uint64_t size = 0;
-        CHECK(_random_file.size(&size).ok());
+        ASSIGN_OR_RETURN(const uint64_t size, _random_file.get_size());
```
Comment (trueeyu) [inline be/src/env/env_memory.cpp:90]: CHECK can't rewrite to this;

File: be/src/env/env_memory.cpp [hunk 5]
```diff
@@ -361,22 +353,20 @@ class EnvMemoryImpl {
...
 
-    Status is_directory(const butil::FilePath& path, bool* is_dir) {
+    StatusOr<bool> is_directory(const butil::FilePath& path) {
         auto inode = get_inode(path);
         if (inode == nullptr) {
             return Status::NotFound(path.value());
         }
-        *is_dir = (inode->type == kDir);
-        return Status::OK();
+        return inode->type == kDir;
```
Comment (trueeyu) [inline be/src/env/env_memory.cpp:361]: invalid rewrite

File: be/src/env/env_memory.cpp [hunk 6]
```diff
@@ -361,22 +353,20 @@ class EnvMemoryImpl {
...
 
-    Status get_file_size(const butil::FilePath& path, uint64_t* size) {
+    StatusOr<uint64_t> get_file_size(const butil::FilePath& path) {
         auto inode = get_inode(path);
         if (inode == nullptr || inode->type != kNormal) {
             return Status::NotFound("not exist or is a directory");
         }
-        *size = inode->data.size();
-        return Status::OK();
+        return inode->data.size();
```
Comment (trueeyu) [inline be/src/env/env_memory.cpp:369]: invalid rewrite

File: be/test/exec/vectorized/hdfs_scanner_test.cpp [hunk 7]
```diff
@@ -59,9 +59,9 @@ void HdfsScannerTest::_create_runtime_state() {
 
 THdfsScanRange* HdfsScannerTest::_create_scan_range(const std::string& file, uint64_t offset, uint64_t length) {
     auto* scan_range = _pool.add(new THdfsScanRange());
-    uint64_t file_size = 0;
-    Status status = Env::Default()->get_file_size(file, &file_size);
-    DCHECK(status.ok()) << status.get_error_msg();
+    StatusOr<uint64_t> status_or = Env::Default()->get_file_size(file);
+    DCHECK(status_or.ok()) << status_or.status().get_error_msg();
+    uint64_t file_size = status_or.value();
```
Comment (sduzh) [inline be/test/exec/vectorized/hdfs_scanner_test.cpp:64]: You can use the macro `ASSIGN_OR_ABORT` defined in `testutil/assert.h`, like this:
```
ASSIGN_OR_ABORT(auto file_size, Env::Default()->get_file_size(file));
```

File: be/test/env/output_stream_wrapper_test.cpp [hunk 8]
```diff
@@ -58,12 +58,13 @@ TEST_F(OutputStreamWrapperTest, test_write) {
     auto rf = *Env::Default()->new_random_access_file(_file->filename());
     std::string buff(21, 0);
     Slice slice(buff);
-    uint64_t size = 0;
-    auto st = rf->size(&size);
-    ASSERT_TRUE(st.ok()) << st;
+
+    const auto status_or = rf->get_size();
+    ASSERT_TRUE(status_or.ok()) << status_or.status();
+    uint64_t size = status_or.value();
```
Comment (sduzh) [inline be/test/env/output_stream_wrapper_test.cpp:64]: You can include `testuti/assert.h` and simplify these code fragment by writing
```
ASSIGN_OR_ABORT(auto size, rf->get_size());
```

**Comment by xlwh** [comment]
Thanks for the review! I have a question: how can I speed up compilation testing in local? It takes me more than three or four hours to compile. My local environment is MacOS 4 core i5. thanks.

File: be/test/env/env_broker_test.cpp [hunk 9]
```diff
@@ -30,17 +30,16 @@ class MockBrokerServer {
...
             return;
-        } else if (!st.ok()) {
+        } else if (!status_or.ok()) {
             response.opStatus.__set_statusCode(TBrokerOperationStatusCode::TARGET_STORAGE_SERVICE_ERROR);
             return;
         }
-        uint64_t size = 0;
-        (void)_env->get_file_size(path, &size);
+        bool is_dir = status_or.value();
+        uint64_t size = _env->get_file_size(path).value();
```
Comment (sduzh) [inline be/test/env/env_broker_test.cpp:42]:
```suggestion
        uint64_t size = is_dir ? _env->get_file_size(path).value() : 0;
``` 
This is why the unit tests failed.

Comment (sduzh) [inline be/test/env/env_broker_test.cpp:42]:
```suggestion
        uint64_t size = is_dir ? _env->get_file_size(path).value() : 0;
``` 
This is why the unit tests failed.

Comment (xlwh) [reply be/test/env/env_broker_test.cpp:42]: Thank you very much! I'll try it later. My local environment is very slow to compile, so it is not easy to debug, any good suggestions?

**Comment by sduzh** [comment]
> Thanks for the review! I have a question: how can I speed up compilation testing in local? It takes me more than three or four hours to compile. My local environment is MacOS 4 core i5. thanks.

Yes, it's quite slow to build StarRocks now.
Here are some tips that may help to improve the compiling speed, but I'm not sure how much improvement they can bring:

- Switch your linker to [mold](https://github.com/rui314/mold) when building StarRocks

- Change the value of the variable `PARALLEL` in `run-ut.sh`
```
PARALLEL=$[$(nproc)/4+1]
```
- Install [gtest-parallel](https://github.com/google/gtest-parallel) and set the environment `GTEST_PARALLEL` before running `run_ut.sh`.
```
export GTEST_PARALLEL=/your/path/to/gtest-parallel
export GTEST_PARALLEL_OPTIONS="-w 50 --print_test_times"
```

**Review by sduzh — APPROVED** [review]

**Comment by sduzh** [comment]
LGTM, @trueeyu @imay PTAL

**Comment by xlwh** [comment]
> > Thanks for the review! I have a question: how can I speed up compilation testing in local? It takes me more than three or four hours to compile. My local environment is MacOS 4 core i5. thanks.
> 
> Yes, it's quite slow to build StarRocks now. Here are some tips that may help to improve the compiling speed, but I'm not sure how much improvement they can bring:
> 
> * Switch your linker to [mold](https://github.com/rui314/mold) when building StarRocks
> * Change the value of the variable `PARALLEL` in `run-ut.sh`
> 
> ```
> PARALLEL=$[$(nproc)/4+1]
> ```
> 
> * Install [gtest-parallel](https://github.com/google/gtest-parallel) and set the environment `GTEST_PARALLEL` before running `run_ut.sh`.
> 
> ```
> export GTEST_PARALLEL=/your/path/to/gtest-parallel
> export GTEST_PARALLEL_OPTIONS="-w 50 --print_test_times"
> ```

Thanks a lot for your suggestion, this looks like it should have some effect

**Review by sduzh — DISMISSED** [review]

File: be/src/env/env_broker.cpp (hunk 1 above)
Comment (trueeyu) [reply be/src/env/env_broker.cpp:478]: if _list_file failed, old implement will set *size，but not set in new implement

File: be/src/env/env_broker.cpp [hunk 10]
```diff
@@ -469,29 +462,27 @@ Status EnvBroker::sync_dir(const std::string& dirname) {
...
-    return st;
+    RETURN_IF_ERROR(_list_file(path, &stat));
+    return stat.size;
 }
 
-Status EnvBroker::get_file_size(const std::string& path, uint64_t* size) {
-    return _get_file_size(path, size);
+StatusOr<uint64_t> EnvBroker::get_file_size(const std::string& path) {
+    return _get_file_size(path);
 }
```
Comment (wyb) [inline be/src/env/env_broker.cpp:483]: These codes can be unified into one function `get_file_size`.

File: be/src/env/env_memory.cpp [hunk 11]
```diff
@@ -91,9 +87,9 @@ class MemorySequentialFile final : public SequentialFile {
     const std::string& filename() const override { return _random_file.filename(); }
 
     Status skip(uint64_t n) override {
-        uint64_t size = 0;
-        CHECK(_random_file.size(&size).ok());
-        _offset = std::min(_offset + n, size);
+        const StatusOr<uint64_t> status_or = _random_file.get_size();
+        CHECK(status_or.ok());
+        _offset = std::min(_offset + n, status_or.value());
```
Comment (wyb) [inline be/src/env/env_memory.cpp:92]:
```suggestion
        ASSIGN_OR_RETURN(auto size, _random_file.get_size());
        _offset = std::min(_offset + n, size);
```

**Review by sduzh — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Pull request merged**